### DISCUSSION TITLE
Uniform Grid thresholding question

### DISCUSSION THREAD

**Discussion by RichardScottOZ** [pyvista/pyvista#1815] [Q&A] [answered]
I am taking an inversion output file - x, y, z and the data (density). Looking at it as a point cloud to see if it makes sense, which it seems to. The explanation suggests this is a uniform grid, so I thought I would make one.

The point cloud thresholds fine, but applying the threshold to the uniform grid reduces the cell/point numbers - but still has data that isn't in the threshold range - although the data array shape makes sense now.

There is something I am not understanding about this perhaps?

I looked at the vtk output.

e.g.

```python
import numpy as np
import pandas as pd
import pyvista as pv

Points = np.loadtxt(r"J:\Richard_Scott\Gravity_Inversion_Constrained SA_Geophysics_Reference_Model_SGrid_ASCIIGravity_Inversion_Constrained__ascii@@", skiprows = 3) #Loading the data only
mesh_points = Points[:,0:3]   ##x, y, z and density

point_cloud = pv.PolyData(mesh_points)
point_cloud['density'] = Points[:,3]

print(Points[:,3].shape)

SARef = pv.UniformGrid()

SARef.dimensions = (333, 378,  29)
SARef.origin = (-180000, 5752000,  -49000)
SARef.spacing = (4000, 4000, 1000)

SARef["density"] = Points[:,3]

SARef.set_active_scalars("density")

SARef["density"].shape

thres = SARef.threshold(value=1.000001, invert=False, preference='point')
print("THRES")
print(thres)

print(SARef['density'].shape)
print(SARef['density'].min())
print(SARef['density'].mean())
print(SARef['density'].max())

print(thres['density'].shape)
print(thres['density'].min())
print(thres['density'].mean())
print(thres['density'].max())

(3650346,)
-99999.0
-40219.61380715547
3.2188799381256104

(2318008,)
-99999.0
-5859.789405599435
3.2188799381256104

```

![image](https://user-images.githubusercontent.com/72196131/142686926-6315fba5-285d-4d61-ab00-64eb913cef1b.png)

![image](https://user-images.githubusercontent.com/72196131/142686938-b1606dec-adde-440d-a4f0-c7847fc736ba.png)

![image](https://user-images.githubusercontent.com/72196131/142686956-e465d071-6594-4739-9098-f2d6f19b43f1.png)

https://drive.google.com/file/d/1Boh48_tjHmM5rwMkJjxDfnkVm0yTdWnR/view?usp=sharing

and I realise the Z is not uniform now, but doesn't matter for this example!
Same behaviour I am not clear on for a Rectilinear Grid, anyway.

A quick other version - voxelising via PVGeo and thresholding works as expected

![image](https://user-images.githubusercontent.com/72196131/142687086-f7451217-28e0-4654-b633-0b683f018a25.png)

Operating on cells though presumably - so the cell/point relationship perhaps I am missing in the open?

```python
vtkThreshold (000001E27F7C0080)
  Debug: Off
  Modified Time: 56623
  Reference Count: 2
  Registered Events: (none)
  Executive: 000001E2254F9B30
  ErrorCode: No error
  Information: 000001E2259B0D60
  AbortExecute: Off
  Progress: 1
  Progress Text: (None)
  Attribute Mode: UseCellData
  Component Mode: UseSelected
  Selected Component: 0
  All Scalars: 0
  Threshold By Upper
  Lower Threshold: 0
  Upper Threshold: 1
  Precision of the output points: 2
  Use Continuous Cell Range: 0


vtkUnstructuredGrid (000001E27F5B7C30)
  Debug: Off
  Modified Time: 56821
  Reference Count: 2
  Registered Events: (none)
  Information: 000001E2259B2110
  Data Released: False
  Global Release Data: Off
  UpdateTime: 56822
  Field Data:
    Debug: Off
    Modified Time: 56769
    Reference Count: 1
    Registered Events: (none)
    Number Of Arrays: 0
    Number Of Components: 0
    Number Of Tuples: 0
  Number Of Points: 2318008
  Number Of Cells: 2203230
  Cell Data:
    Debug: Off
    Modified Time: 56794
    Reference Count: 1
    Registered Events: 
      Registered Observers:
        vtkObserver (000001E225B81880)
          Event: 33
          EventName: ModifiedEvent
          Command: 000001E2259B26B0
          Priority: 0
          Tag: 1
    Number Of Arrays: 0
    Number Of Components: 0
    Number Of Tuples: 0
    Copy Tuple Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Interpolate Flags: ( 1 1 1 1 1 1 0 1 1 1 1 )
    Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Scalars: (none)
    Vectors: (none)
    Normals: (none)
    TCoords: (none)
    Tensors: (none)
    GlobalIds: (none)
    PedigreeIds: (none)
    EdgeFlag: (none)
    Tangents: (none)
    RationalWeights: (none)
    HigherOrderDegrees: (none)
  Point Data:
    Debug: Off
    Modified Time: 56792
    Reference Count: 1
    Registered Events: 
      Registered Observers:
        vtkObserver (000001E225B81820)
          Event: 33
          EventName: ModifiedEvent
          Command: 000001E2259B26B0
          Priority: 0
          Tag: 1
    Number Of Arrays: 1
    Array 0 name = density
    Number Of Components: 1
    Number Of Tuples: 2318008
    Copy Tuple Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Interpolate Flags: ( 1 1 1 1 1 1 0 1 1 1 1 )
    Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Scalars: 
      Debug: Off
      Modified Time: 56789
      Reference Count: 2
      Registered Events: (none)
      Name: density
      Data type: double
      Size: 2318008
      MaxId: 2318007
      NumberOfComponents: 1
      Information: 0000000000000000
      Name: density
      Number Of Components: 1
      Number Of Tuples: 2318008
      Size: 2318008
      MaxId: 2318007
      LookupTable: (none)
    Vectors: (none)
    Normals: (none)
    TCoords: (none)
    Tensors: (none)
    GlobalIds: (none)
    PedigreeIds: (none)
    EdgeFlag: (none)
    Tangents: (none)
    RationalWeights: (none)
    HigherOrderDegrees: (none)
  Bounds: 
    Xmin,Xmax: (-180000, 1.148e+06)
    Ymin,Ymax: (5.752e+06, 7.26e+06)
    Zmin,Zmax: (-49000, 0)
  Compute Time: 56873
  Editable: false
  Number Of Points: 2318008
  Point Coordinates: 000001E225D46D50
  PointLocator: 0000000000000000
  CellLocator: 0000000000000000
  Number Of Pieces: 1
  Piece: 0
  Ghost Level: 0

```

Making a grid of course much quicker than the voxel process!

**Comment by banesullivan** [comment, marked as answer]
The threshold filter is a CELL-wise operation. This means if you have point arrays (which you have in your case), then any cells that contain those points will remain in the result and subsequently those cells other points which may not meet your thresholding criteria. 

Here is a more minimal example:

```py
from pyvista import examples
import pyvista as pv

mesh = examples.load_uniform()
pc = pv.PolyData(mesh.points)
pc.point_data.update(mesh.point_data)

threshed = mesh.threshold(200, scalars='Spatial Point Data')
good_points = pc.threshold(200, scalars='Spatial Point Data')

bad_point_ids = np.argwhere(threshed['Spatial Point Data'] < 200).ravel()
bad_points = pv.PolyData(threshed.points[bad_point_ids])

bad_cell = threshed.extract_cells([0])

p = pv.Plotter(notebook=0)
p.add_mesh(threshed, opacity=.9, show_edges=True, clim=[0, 700])
p.add_mesh(good_points, color='green', point_size=40, render_points_as_spheres=True)
p.add_mesh(bad_points, color='red', point_size=25)
p.add_mesh(bad_cell, color='pink', opacity=0.75)

for i, pt in enumerate(bad_cell.points):
    if bad_cell['Spatial Point Data'][i] < 200:
        label = 'Bad'
    else:
        label = 'Good'
    p.add_point_labels(pt, [label, ], point_size=20, font_size=36)
    
p.camera_position = [
 (12.04307118629237, 4.946603092408221, 0.42877398300085057),
 (6.5398207910337876, 8.478480473518601, 4.025742801287316),
 (-0.575598431297095, -0.0668633476072612, -0.8149943181612213)]
p.show()
```


<img width="624" alt="Screen Shot 2021-11-25 at 11 39 06 PM" src="https://user-images.githubusercontent.com/22067021/143537880-bc4e8599-6967-49a5-aeb8-9fb3556725d5.png">

What I'm trying to demonstrate is that one good point in a cell will cause that cell to be extracted and thus all the other points at its corners which could be bad.

There are some ways around this but they can get fairly involved or have a computational cost

**Reply by banesullivan** [reply]
Furthermore, it isn't as simple as just removing the cells with bad points as removing a cell with one bad point might remove some good points):

```py
threshed['cell_ids'] = np.arange(threshed.n_cells)
bad_cells = threshed.extract_points(bad_point_ids)['cell_ids']
cleaned = threshed.copy().remove_cells(bad_cells)

p = pv.Plotter(notebook=0)
p.add_mesh(cleaned.extract_all_edges(), color='green', line_width=10)
p.add_mesh(threshed, color='red', opacity=0.5)
p.add_mesh(good_points, color='yellow', point_size=10)
p.show()
```

<img width="624" alt="Screen Shot 2021-11-25 at 11 48 20 PM" src="https://user-images.githubusercontent.com/22067021/143538852-013a2363-1665-46b5-8727-345f2649f382.png">

**Comment by RichardScottOZ** [comment]
Thanks - what is the computational cost with comparison to voxelisation?

**Reply by banesullivan** [reply]
That voxelization filter in PVGeo isn't all that efficient... I wrote that long before I mastered the ways of the NumPy and the VTK. The voxelization filter is actually converting your data to be CELL data as a voxel is created at every point (the center of the created voxels are the input points). Honestly.... if that filter in PVGeo works for you, use it! but do take note of the fact that it is converting your points to CELL data

IMO, since your dealing with density inversion code you might actually want your density array to be CELL data on the UniformGrid and not point data. Right now, you are generating your UniformGrid so that it's points line up with the points in the point cloud you have. But if I'm assuming correctly, your inversion code likely uses cells/regions of space to perform the modeling and the output point cloud you have is likely just a representation of the centers of those regions -- I'd look into how that point cloud is generated a bit more.

What you may want to do is "shift" the uniform grid such that the cell centers line up with the point cloud you have. This would mean the `n_cells` in the `UniformGrid` would equal the `n_points` in the point cloud. This would be insanely more efficient than PVGeo's voxelization filter.

**Comment by RichardScottOZ** [comment]
Thanks very much - that is pretty much along the lines of what I was considering, given cell data conversions.

And similarly for Rectilinear grid with varying Z depth.

**Comment by RichardScottOZ** [comment]
![image](https://user-images.githubusercontent.com/72196131/143543975-f6b729af-fda8-4464-a433-6cba9fad7e4f.png)

**Comment by RichardScottOZ** [comment]
![image](https://user-images.githubusercontent.com/72196131/143664177-543ceb24-9f2e-460d-807a-08dbdc3e12f8.png)
